File: C++/Sum Root to Leaf Numbers.cpp
#include<iostream>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

void dfs(TreeNode *root,int sum,int &res)
    {
        if(root==NULL) return;
        else if(root->left==NULL&&root->right==NULL)
            {
                res=res+sum+root->val;
            }
        else if(root->left==NULL&&root->right!=NULL)
            {
                //res=res+sum;
                dfs(root->right,(sum+root->val)*10,res);
            }
        else if(root->left!=NULL&&root->right==NULL)
            {
                // res=res+sum;
                dfs(root->left,(sum+root->val)*10,res);
            }
        else if(root->left!=NULL&&root->right!=NULL)
            {
                dfs(root->left,(sum+root->val)*10,res);
                dfs(root->right,(sum+root->val)*10,res);
            }
    }

int sumNumbers(TreeNode *root)
    {
        if(root==NULL) return 0;
        int cnt=0;
        dfs(root,0,cnt);
        return cnt;
    }

File: Python/Sum-Root-to-Leaf-Numbers.py
#!/usr/bin/env python

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return an integer
    def sumroot(self, root, s):
        if root == None:
            return 0
        elif root.left == None and root.right == None:
            return s * 10 + root.val
        else:
            return self.sumroot(root.left, s * 10 + root.val) + self.sumroot(root.right, s * 10 + root.val)
    def sumNumbers(self, root):
        return self.sumroot(root, 0)